### PR TITLE
feat: Add support for Ollama as a local LLM via litellm

### PR DESCRIPTION
**Pull request by cguedes**
related to #58 and #59

This PR adds support for using Ollama as a local LLM via `litellm`.

To use refstudio with [Ollama](https://ollama.ai/), you'll need to ensure that both the Ollama server and refstudio are running on your local machine at the same time.
1. Download and install Ollama
2. Launch the Ollama server locally: `ollama serve`
3. Verify the Ollama server is working: `ollama run llama2 "are you running?"`
4. Launch refstudio
5. In refstudio > settings, change your `chat_model` setting to `llama2`
_____________
This is an experimental PR trying to integrate [LiteLLM](https://github.com/BerriAI/litellm) with RefStudio with a local LLM using [Ollama](https://github.com/jmorganca/ollama).

**Notes:**
* ~🤔 I'm not sure if the async/await should be used here. I needed to use that (and the unwrap for the local LLM. And then implemented the same call (`litellm.acompletion`) for the OpenAI~
* ~💥 I haven't fixed unit test. They need to be adjusted to use the new async methods~
* ⚠️ We still need to find a way to configure the open AI api key. I think that the litellm would fallback to the env var, something we don't want
* ℹ️ I've made simple to toggle between `gpt-3.5-turbo` and `llama2` in the settings. I think we still want to keep the input field, and we don't have a dropdown+input form element (the one that allow to select an option but also write any text we want).
* ~💥 The desktop can't start with LiteLLM! I've `yarn python` and then running `yarn tauri:dev:debug` would show the following~

To run this example, first install Ollama (with llama2) and start it:

```bash
ollama serve
ollama run llama2 'are you running?'
```

To test inside Refstudio (in the rewrite) you just need to configure the correct model in the settings before each rewrite. You can use `llama2` (local Ollama) or `gpt-3.5-turbo` (for OpenAI - you need the API Key correctly configured).

image

image

### CODE REVIEW

**Review by sehyod — COMMENTED** [review]
You should update the comment of the `Rewriter` class (lines 164 to 177 in rewrite.py)

**Review by sehyod — COMMENTED** [review]
The frontend part looks great!
The backend part does not work when running the desktop app (the server cannot start so the app doesn't start). And as discussed offline, I don't understand what API key we are using, but even with an invalid key in my settings, the requests are successful

**Comment by gjreda** [comment]
Chatted with @cguedes offline and I'll plan to pick up this PR on Monday and work through the remaining backend issues.

**Comment by ishaan-jaff** [comment]
@cguedes i'm the maintainer of litellm sorry for the errors and poor experience you encountered:
Looks like there are two main action items to improve litellm here (please let me know if there are more)

- yarn python fails with litellm looks like `get_encoding fails` and then leads to a bunch of errors 
- Don't fallback to .env for checking keys @cguedes how would you pass keys to litellm instead ?

**Comment by gjreda** [comment]
@cguedes @sehyod IMO all that's left here is:
1. Rename the `openai` related backend settings to be more generic. I'm thinking we'll have:
- `provider` - The LLM service/company to use (e.g. `openai`, `anthropic`, `ollama`)
- `api_key` - The API key for the specified provider (`None` for ollama)
- `model` - No need to call this `chat_model` since it is used by all AI interactions at this point
2. Update the frontend settings accordingly.

Also not sure why the prettier check is failing since it runs fine locally and on the pre-commit.

File: .github/workflows/test-sidecar-binary.yml
Comment (gjreda) [inline .github/workflows/test-sidecar-binary.yml:1]: Feels like we should always be testing this when python files are changed

**Comment by cguedes** [comment]
@gjreda the client side code is ready. This is the updated settings panel.

image

**Review by gjreda — APPROVED** [review]

File: src/features/ai/settings/AiSettingsPane.tsx
Comment (sehyod) [inline src/features/ai/settings/AiSettingsPane.tsx:102]: My understanding of the backend code is that the model is overridden, to use 'llama2' when the provider is ollama. Maybe we should reflect that in the frontend by disabling the field and displaying 'llama' when selecting ollama provider?

**Review by sehyod — COMMENTED** [review]
I have tried the branch locally and I have 2 comments:
 - I don't think we are properly using the API key from the settings: even with an invalid key (or without any key at all) I am still able to rewrite text
 - I haven't been able to use ollama provider: I have installed everything, checked the server was working with `ollama run llama2 "are you running?"` and I get an error when trying to rewrite my selection (`RetryError[<Future at 0x10f557210 state=finished raised TypeError>]`)

**Comment by cguedes** [comment]
> I have tried the branch locally and I have 2 comments:
> 
> * I don't think we are properly using the API key from the settings: even with an invalid key (or without any key at all) I am still able to rewrite text
> * I haven't been able to use ollama provider: I have installed everything, checked the server was working with `ollama run llama2 "are you running?"` and I get an error when trying to rewrite my selection (`RetryError[<Future at 0x10f557210 state=finished raised TypeError>]`)

@sehyod thanks for the feedback.  

1. The library fallback to an env var. Will check that behaviour again
2. That is fixed that in my latest commit

**Comment by sehyod** [comment]
> 1. The library fallback to an env var. Will check that behaviour again
> 2. That is fixed that in my latest commit

1. I don't have any .env file so it doesn't explain how it works, and I think it should not fallback to an env var if the user provides a key through the app, even if the key is not valid
2. Thanks for the fix! I confirm it is working now 🎉

File: src/features/ai/settings/AiSettingsPane.tsx
Comment (cguedes) [reply src/features/ai/settings/AiSettingsPane.tsx:102]: I will make the llama2 default when we select Ollama and same for OpenAI. They can use other values like gpt4 or other ollama models.

File: python/sidecar/ai/chat.py
Comment (sehyod) [inline python/sidecar/ai/chat.py:31]: I think this should be `os.environ["OPENAI_API_KEY"] = user_settings.api_key`: openai.api_key doesn't seem to be used, which is why I was getting a response even when my key was invalid

Comment (sehyod) [reply python/sidecar/ai/chat.py:31]: After more investigation, it appears that when `openai.api_key` is invalid, liteLLM falls back to the `OPENAI_API_KEY` env var. I have no clue why my env var is correct (I don't have an .env file) but this is what was happening.
This behavior is a bit confusing, I think we should not fallback to anything if there is an api key set in the app. Carlos told me that the plan was to send the api_key in the request, instead of using `openai.api_key`, maybe this would prevent the fallback. In any case, this is a very edge case because the user has to have an `OPENAI_API_KEY` env var for the fallback to happen, so I don't think it should block this PR, but we have to keep it in mind

**Review by sehyod — APPROVED** [review]

**Pull request merged**
